File: course_level1_py/course_level1_py/number_publisher.py
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from example_interfaces.msg import Int64

class NumberPublisherNode(Node):
    def __init__(self):
        super().__init__("number_publisher")
        self.declare_parameter("number", 2)
        self.declare_parameter("timer_period", 1.0)
        self.number_ = self.get_parameter("number").value
        self.timer_period_ = self.get_parameter("timer_period").value
        self.add_post_set_parameters_callback(self.parameters_callback)

        # If you add a '/' before 'number', it will make it unchangeable
        # by namespaces and remain as /number
        self.pub_ = self.create_publisher(Int64, "number", 10)
        self.timer_ = self.create_timer(self.timer_period_, self.publish_number)
        self.get_logger().info("Number Publisher has been started!")

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.pub_.publish(msg)

    # Used if we want to change parameter during runtime 
    # Use with caution (for example for timer_period would 
    # have to do .cancel and then relaunch with right value)
    def parameters_callback(self, params: list[Parameter]): 
        for param in params:
            if param.name == "number":
                self.number_ = param.value

def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()
    
    
if __name__ == "__main__":
    main()

File: course_level1/course_level1_cpp/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;

class NumberNode : public rclcpp::Node
{
public:
    NumberNode() : Node("number_publisher")
    {
        this->declare_parameter("number", 2);
        this->declare_parameter("timer_period", 1.0);
        number_ = this->get_parameter("number").as_int();
        double timer_period_ = this->get_parameter("timer_period").as_double();

        param_callback_handle_ = this->add_post_set_parameters_callback(
            std::bind(&NumberNode::parametersCallback, this, _1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period_), 
            std::bind(&NumberNode::publish_number, this));
        RCLCPP_INFO(this->get_logger(), "Number Publisher has been started.");

    }
private:
    void publish_number()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        publisher_->publish(msg);
    }

    // Used if we want to change parameter during runtime 
    // Use with caution (for example for timer_period would 
    // have to do .cancel and then relaunch with right value)
    void parametersCallback(const std::vector<rclcpp::Parameter> & parameters)
    {
        for (const auto &param: parameters) {
            if (param.get_name() == "number") {
                number_ = param.as_int();
            }
        }
    }
    int number_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    PostSetParametersCallbackHandle::SharedPtr param_callback_handle_;
};


int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}